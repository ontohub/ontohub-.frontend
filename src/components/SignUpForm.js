import React, { Component } from 'react'
import zxcvbn from 'zxcvbn'
import { validate } from '../helpers/validations'
import { Button, Form, Label, Progress } from 'semantic-ui-react'
import _ from 'lodash'

const PasswordStrengthBar = ({ score }) =>
  <Progress
    attached="bottom"
    title={`Password strength: ${[
      'Very weak',
      'Weak',
      'Okay',
      'Strong',
      'Very strong'
    ][score]}`}
    style={{ marginTop: -3 }}
    percent={score * 25}
    color={['red', 'orange', 'yellow', 'olive', 'green'][score]}
  />

export class SignUpForm extends Component {
  constructor(props) {
    super(props)
    this.fields = {}
    this.state = {
      passwordScore: 0,
      errors: {}
    }
  }
  componentDidMount() {
    this._isMounted = true
  }
  componentWillUnmount() {
    this._isMounted = false
  }
  fieldValues() {
    return _.mapValues(this.fields, (f) => f && f.value)
  }
  onSubmit(e) {
    e.preventDefault()
    let fieldValues = this.fieldValues(),
        validationErrors = validate(this.props.validations, fieldValues)
    return validationErrors.then((errors) => {
      this.setState({
        errors: errors
      })

      if (Object.keys(errors).length === 0) {
        return this.props
          .onSubmit(fieldValues)
          .then(this.props.onSuccess, this.props.onError)
      } else {
        return this.props.onError()
      }
    })
  }
  validateField(fieldName) {
    return _.debounce(
      (e) =>
        validate(
          this.props.validations,
          this.fieldValues(),
          fieldName
        ).then((errors) => {
          if (this._isMounted) {
            this.setState((state) => ({
              ...state,
              errors: { ...state.errors, ...errors }
            }))
          }
        }),
      300
    )
  }
  render() {
    return (
      <Form onSubmit={this.onSubmit.bind(this)}>
        <Form.Group widths="equal">
          <Form.Field error={!!this.state.errors.username}>
            <label htmlFor="sign-up-username">Username</label>
            <input
              ref={(input) => (this.fields.username = input)}
              onKeyPress={this.validateField('username')}
              onBlur={this.validateField('username')}
              placeholder="Username"
              id="sign-up-username"
            />
            {this.state.errors.username &&
              <Label pointing color="red">
                {this.state.errors.username.map((e, i) =>
                  <div key={i}>{e}</div>
                )}
              </Label>}
          </Form.Field>
          <Form.Field error={!!this.state.errors.email}>
            <label htmlFor="sign-up-email">Email</label>
            <input
              ref={(input) => (this.fields.email = input)}
              onKeyPress={this.validateField('email')}
              onBlur={this.validateField('email')}
              placeholder="Email"
              id="sign-up-email"
            />
            {this.state.errors.email &&
              <Label pointing color="red">
                {this.state.errors.email.map((e, i) => <div key={i}>{e}</div>)}
              </Label>}
          </Form.Field>
        </Form.Group>
        <Form.Group widths="equal">
          <Form.Field error={!!this.state.errors.password}>
            <label htmlFor="sign-up-password">Password</label>
            <input
              type="password"
              ref={(input) => (this.fields.password = input)}
              onKeyPress={this.validateField(['password', 'passwordConfirm'])}
              onBlur={this.validateField(['password', 'passwordConfirm'])}
              onChange={(e) => {
                let score = zxcvbn(e.target.value).score
                this.setState({
                  passwordScore: score
                })
              }}
              placeholder="Password"
              id="sign-up-password"
            />
            <PasswordStrengthBar score={this.state.passwordScore} />
            {this.state.errors.password &&
              <Label pointing color="red">
                {this.state.errors.password.map((e, i) =>
                  <div key={i}>{e}</div>
                )}
              </Label>}
          </Form.Field>
          <Form.Field error={!!this.state.errors.passwordConfirm}>
            <label htmlFor="sign-up-password-confirm">Confirm Password</label>
            <input
              type="password"
              ref={(input) => (this.fields.passwordConfirm = input)}
              onKeyPress={this.validateField('passwordConfirm')}
              onBlur={this.validateField('passwordConfirm')}
              placeholder="Confirm Password"
              id="sign-up-password-confirm"
            />
            {this.state.errors.passwordConfirm &&
              <Label pointing color="red">
                {this.state.errors.passwordConfirm.map((e, i) =>
                  <div key={i}>{e}</div>
                )}
              </Label>}
          </Form.Field>
        </Form.Group>
        <Button
          basic
          type="submit"
          icon="add user"
          content="Create account"
          floated="right"
          secondary
        />
      </Form>
    )
  }
}

export default SignUpForm
