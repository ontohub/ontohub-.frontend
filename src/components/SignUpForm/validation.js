import { Client, userQuery } from "config/apollo";
import { get, isEmpty, filter, pick } from "lodash";
import debounce from "debounce-promise";

export const isUsernameAvailable = id =>
  Client.query({ query: userQuery, variables: { id: id } }).then(
    data => !get(data, "data.organizationalUnit", null)
  );

const isUsernameAvailableDebounced = debounce(isUsernameAvailable, 500, {
  leading: true
});

export const formHasError = errors =>
  !isEmpty(
    filter(pick(errors, ["email", "name", "password", "confirmPassword"]))
  );

export const addError = (condition, errors, field, error) => {
  if (condition) {
    if (!errors[field]) {
      errors[field] = [];
    }
    errors[field].push(error);
  }
};

export const validate = values => {
  return import("zxcvbn")
    .then(fn => {
      let errors = {};
      let scorePassword = password => fn(password).score;

      errors.passwordScore = scorePassword(values.password);

      addError(
        !values.password || values.password.length < 10,
        errors,
        "password",
        "Password must be at least 10 characters long"
      );
      addError(
        values.password !== values.confirmPassword,
        errors,
        "confirmPassword",
        "Passwords must match"
      );
      addError(
        !/^[^@]+@[^@]+$/.test(values.email),
        errors,
        "email",
        "Email is invalid"
      );
      addError(
        !values.name || values.name.length < 3,
        errors,
        "name",
        "Username must be at least 3 characters long"
      );
      addError(
        !/[a-z0-9][a-z0-9\-_]*[a-z0-9]/.test(values.name),
        errors,
        "name",
        "Username must start and end with a lower case letter " +
          "or number, and only contain lower case " +
          'letters, numbers, "-" and "_"'
      );
      return errors;
    })
    .then(errors => {
      /* istanbul ignore else */
      if (!errors.name) {
        return isUsernameAvailableDebounced(values.name).then(
          available => {
            addError(!available, errors, "name", "Username is already taken");
            /* istanbul ignore else */
            if (formHasError(errors)) {
              throw errors;
            } else {
              return errors;
            }
          },
          err => {
            addError(
              true,
              errors,
              "name",
              "Unable to check username availability. Please check your network connection."
            );
            throw errors;
          }
        );
      } else if (formHasError(errors)) {
        throw errors;
      } else {
        return errors;
      }
    });
};
