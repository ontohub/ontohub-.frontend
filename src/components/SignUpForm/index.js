import React from "react";
import Field from "./Field";
import PasswordStrengthBar from "./PasswordStrengthBar";
import withRefs from "./withRefs";
import { compose } from "recompose";
import { Button, Form } from "semantic-ui-react";
import { Formik } from "formik";
import { validate } from "./validation";
import { setServerErrors } from "lib/validation";
import ReCAPTCHA from "react-google-recaptcha";
const grecaptchaSiteKey = process.env.REACT_APP_GRECAPTCHA_SITE_KEY;

export const SignUpForm = ({
  captcha: Captcha,
  errors,
  onBlur,
  onChange,
  onSubmit,
  submitEnabled = true,
  touched,
  values
}) => (
  <Form onSubmit={submitEnabled ? onSubmit : undefined}>
    {Captcha || null}
    <Form.Group widths="equal">
      <Field
        errors={errors.name}
        fieldId="sign-up-username"
        label="Username"
        name="name"
        onBlur={onBlur}
        onChange={onChange}
        touched={touched.name}
        value={values.name}
      />
      <Field
        errors={errors.email}
        fieldId="sign-up-email"
        label="Email"
        name="email"
        onBlur={onBlur}
        onChange={onChange}
        touched={touched.email}
        value={values.email}
      />
    </Form.Group>
    <Form.Group widths="equal">
      <Field
        bar={<PasswordStrengthBar score={errors.passwordScore || 0} />}
        errors={errors.password}
        fieldId="sign-up-password"
        label="Password"
        name="password"
        onBlur={onBlur}
        onChange={onChange}
        touched={touched.password}
        type="password"
        value={values.password}
      />
      <Field
        errors={errors.confirmPassword}
        fieldId="sign-up-password-confirm"
        label="Password confirmation"
        name="confirmPassword"
        onBlur={onBlur}
        onChange={onChange}
        touched={touched.confirmPassword}
        type="password"
        value={values.confirmPassword}
      />
    </Form.Group>
    <Button
      basic
      type="submit"
      icon="add user"
      content="Create account"
      floated="right"
      secondary
      disabled={!submitEnabled}
    />
  </Form>
);

const initialValues = {
  name: "",
  email: "",
  password: "",
  confirmPassword: ""
};

/* istanbul ignore next */
const FormikSignUpForm = ({
  enableCaptcha,
  onError,
  onSubmit,
  refs,
  registerCaptcha,
  setValues
}) => (
  <Formik
    initialValues={initialValues}
    onSubmit={(values, { setErrors, setSubmitting }) => {
      if (refs.captcha && enableCaptcha) {
        refs.captcha.reset();
        refs.captcha.execute();
      } else {
        onSubmit({ ...values, captcha: "skip" }).catch(
          setServerErrors(setErrors, setSubmitting, onError)
        );
      }
    }}
    render={({
      errors,
      handleBlur,
      handleChange,
      handleSubmit,
      setErrors,
      setFieldTouched,
      setSubmitting,
      touched,
      values
    }) => (
      <SignUpForm
        captcha={
          enableCaptcha &&
          initialValues !== values && (
            <ReCAPTCHA
              sitekey={grecaptchaSiteKey}
              ref={registerCaptcha}
              size="invisible"
              badge="bottomleft"
              onChange={captcha => {
                onSubmit({ ...refs.values, captcha }).catch(
                  setServerErrors(setErrors, setSubmitting, onError)
                );
              }}
            />
          )
        }
        errors={errors}
        onBlur={handleBlur}
        onChange={(e, ...args) => {
          const fieldName = e.target.name;
          const fieldValue = e.target.value;
          setFieldTouched(fieldName, true);
          setValues({ ...refs.values, [fieldName]: fieldValue });
          return handleChange(e, ...args);
        }}
        onSubmit={handleSubmit}
        submitEnabled={values !== initialValues}
        touched={touched}
        values={values}
      />
    )}
    validate={validate}
  />
);

const FormikSignUpFormWithRefs = compose(
  withRefs({ captcha: "registerCaptcha", values: "setValues" })
)(FormikSignUpForm);

export default FormikSignUpFormWithRefs;
