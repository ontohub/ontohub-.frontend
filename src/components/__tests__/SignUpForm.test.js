import React from 'react'
import { shallow, mount } from 'enzyme'
import _ from 'lodash'
import SignUpForm from '../SignUpForm'

jest.genMockFromModule('zxcvbn')

describe('SignUpForm', () => {
  let wrapper

  beforeAll(() => {
    wrapper = shallow(<SignUpForm />)
  })
  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('onSubmit', () => {
    describe('error', () => {
      let onError, onSubmit, validations

      beforeAll(() => {
        onError = jest.fn()
        onSubmit = () => Promise.reject()
        validations = {
          username: ['is invalid', 'is too short'],
          email: ['is invalid'],
          password: ['is invalid'],
          passwordConfirm: ['is invalid']
        }
        wrapper = mount(
          <SignUpForm
            validations={validations}
            onSubmit={onSubmit}
            onError={onError}
            enableCaptcha={false}
          />
        )
        wrapper.instance().onSubmit({ preventDefault: jest.fn() })
      })

      it('matches the snapshot', () => {
        wrapper.setState({ error: true })
        expect(wrapper).toMatchSnapshot()
      })

      it('calls the error callback', () => {
        expect(onError.mock.calls.length).toBe(1)
      })
    })

    describe('success', () => {
      let onSuccess, onSubmit, validations

      beforeAll(() => {
        onSuccess = jest.fn()
        onSubmit = jest.fn(() => Promise.resolve())
        validations = {}
        wrapper = mount(
          <SignUpForm
            validations={validations}
            onSubmit={onSubmit}
            onSuccess={onSuccess}
            enableCaptcha={false}
          />
        )
        _.each(['username', 'email', 'password'], v => {
          wrapper.instance().fields[v].value = v
        })
        wrapper.instance().onSubmit({ preventDefault: jest.fn() })
      })

      it('has not yet loaded recaptcha', () => {
        expect(wrapper.instance().state.captchaLoaded).toBeFalsy()
      })

      it('loads recaptcha onChange', () => {
        wrapper.find('form').simulate('change')
        expect(wrapper.instance().state.captchaLoaded).toBe(true)
      })

      it('calls the submit callback once', () => {
        expect(onSubmit.mock.calls.length).toBe(1)
      })

      it('calls the submit callback with the right arguments', () => {
        // 'skip' is used for the captcha when captcha verification is disabled
        expect(onSubmit.mock.calls[0]).toEqual([
          'username',
          'email',
          'password',
          'skip'
        ])
      })
    })
  })

  describe('Password score', () => {
    beforeAll(() => {
      wrapper = mount(<SignUpForm />)
    })

    it('is calculated when the password is changed', () => {
      let oldScore = wrapper.state('passwordScore')
      wrapper.find('#sign-up-password').simulate('change', {
        target: { value: 'areallylongpasswordthatshouldgetareallyhighscore' },
        no_validation: true
      })
      expect(wrapper.state('passwordScore')).not.toEqual(oldScore)
    })
  })
})
