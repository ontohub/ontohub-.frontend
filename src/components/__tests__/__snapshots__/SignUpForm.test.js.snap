// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm matches the snapshot 1`] = `
<Form
  as="form"
  onChange={[Function]}
  onSubmit={[Function]}
>
  <FormGroup
    widths="equal"
  >
    <FormField
      error={false}
    >
      <label
        htmlFor="sign-up-username"
      >
        Username
      </label>
      <input
        id="sign-up-username"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Username"
      />
    </FormField>
    <FormField
      error={false}
    >
      <label
        htmlFor="sign-up-email"
      >
        Email
      </label>
      <input
        id="sign-up-email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
      />
    </FormField>
  </FormGroup>
  <FormGroup
    widths="equal"
  >
    <FormField
      error={false}
    >
      <label
        htmlFor="sign-up-password"
      >
        Password
      </label>
      <input
        id="sign-up-password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
      <PasswordStrengthBar
        score={0}
      />
    </FormField>
    <FormField
      error={false}
    >
      <label
        htmlFor="sign-up-password-confirm"
      >
        Confirm Password
      </label>
      <input
        id="sign-up-password-confirm"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Confirm Password"
        type="password"
      />
    </FormField>
  </FormGroup>
  <Button
    as="button"
    basic={true}
    content="Create account"
    floated="right"
    icon="add user"
    secondary={true}
    type="submit"
  />
</Form>
`;

exports[`SignUpForm onSubmit error matches the snapshot 1`] = `
<SignUpForm
  enableCaptcha={false}
  onError={[Function]}
  onSubmit={[Function]}
  validations={
    Object {
      "email": Array [
        "is invalid",
      ],
      "password": Array [
        "is invalid",
      ],
      "passwordConfirm": Array [
        "is invalid",
      ],
      "username": Array [
        "is invalid",
        "is too short",
      ],
    }
  }
>
  <Form
    as="form"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <form
      className="ui form"
      onChange={[Function]}
      onSubmit={[Function]}
    >
      <FormGroup
        widths="equal"
      >
        <div
          className="equal width fields"
        >
          <FormField
            error={true}
          >
            <div
              className="error field"
            >
              <label
                htmlFor="sign-up-username"
              >
                Username
              </label>
              <input
                id="sign-up-username"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Username"
              />
              <Label
                color="red"
                pointing={true}
              >
                <div
                  className="ui red pointing label"
                  onClick={[Function]}
                >
                  <div>
                    Username is invalid
                  </div>
                  <div>
                    Username is too short
                  </div>
                </div>
              </Label>
            </div>
          </FormField>
          <FormField
            error={true}
          >
            <div
              className="error field"
            >
              <label
                htmlFor="sign-up-email"
              >
                Email
              </label>
              <input
                id="sign-up-email"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Email"
              />
              <Label
                color="red"
                pointing={true}
              >
                <div
                  className="ui red pointing label"
                  onClick={[Function]}
                >
                  <div>
                    Email is invalid
                  </div>
                </div>
              </Label>
            </div>
          </FormField>
        </div>
      </FormGroup>
      <FormGroup
        widths="equal"
      >
        <div
          className="equal width fields"
        >
          <FormField
            error={true}
          >
            <div
              className="error field"
            >
              <label
                htmlFor="sign-up-password"
              >
                Password
              </label>
              <input
                id="sign-up-password"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Password"
                type="password"
              />
              <PasswordStrengthBar
                score={0}
              >
                <Progress
                  attached="bottom"
                  color="red"
                  percent={0}
                  style={
                    Object {
                      "marginTop": -3,
                    }
                  }
                  title="Password strength: Very weak"
                >
                  <div
                    className="ui red bottom attached progress"
                    data-percent={0}
                    style={
                      Object {
                        "marginTop": -3,
                      }
                    }
                    title="Password strength: Very weak"
                  >
                    <div
                      className="bar"
                      style={
                        Object {
                          "width": "0%",
                        }
                      }
                    />
                  </div>
                </Progress>
              </PasswordStrengthBar>
              <Label
                color="red"
                pointing={true}
              >
                <div
                  className="ui red pointing label"
                  onClick={[Function]}
                >
                  <div>
                    Password is invalid
                  </div>
                </div>
              </Label>
            </div>
          </FormField>
          <FormField
            error={true}
          >
            <div
              className="error field"
            >
              <label
                htmlFor="sign-up-password-confirm"
              >
                Confirm Password
              </label>
              <input
                id="sign-up-password-confirm"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Confirm Password"
                type="password"
              />
              <Label
                color="red"
                pointing={true}
              >
                <div
                  className="ui red pointing label"
                  onClick={[Function]}
                >
                  <div>
                    PasswordConfirm is invalid
                  </div>
                </div>
              </Label>
            </div>
          </FormField>
        </div>
      </FormGroup>
      <Button
        as="button"
        basic={true}
        content="Create account"
        floated="right"
        icon="add user"
        secondary={true}
        type="submit"
      >
        <button
          className="ui basic secondary right floated button"
          onClick={[Function]}
          role="button"
          type="submit"
        >
          <Icon
            as="i"
            name="add user"
          >
            <i
              aria-hidden="true"
              className="add user icon"
            />
          </Icon>
          Create account
        </button>
      </Button>
    </form>
  </Form>
</SignUpForm>
`;
